Add ScriptEngineSelector to select script engine

danielb987 opened the pull request:
This PR solves issue #10555 but it doesn't complete it. I haven't converted Logix for example.

The PR has the classes `ScriptEngineSelector`, `ScriptEngineSelectorXml` and `ScriptEngineSelectorSwing`.

This makes it easy for Logix, LogixNG and others to let the user select the script engine, without knowing anything about the inner details of ScriptEngineSelector.

Any comments are welcome.

danielb987 commented:
@bobjacobsen 
Can you look at this PR? The old style of running scripts works fine. The new style, using `ScriptEngineSelector`, works in **one** test. But when I use it in several tests, it fails. See `jmri.script.ScriptEngineSelectorTest` which demonstrates this.

The three old style tests in this class runs fine. But only one of the new style tests succeeds. The other two fails.

If you want to continue working on this PR on your own and create your own PR from it, feel free to do so.

bobjacobsen commented:
Am looking at the failures in `ScriptEngineSelectorTest`. I don't see why they're happening yet, which is a bit frustrating. The script is definitely running and changing the turnout, but somehow that's only in the context of the first run when it returns.

danielb987 commented:
> Am looking at the failures in `ScriptEngineSelectorTest`. I don't see why they're happening yet, which is a bit frustrating. The script is definitely running and changing the turnout, but somehow that's only in the context of the first run when it returns.

Thank you! That solved it.

The old style calls `JmriScriptEngineManager.getDefault()` which in turn call `InstanceManager.getDefault(JmriScriptEngineManager.class)`. My version of `ScriptEngineSelector` stores all the engines in a static list which never gets cleared when the InstanceManager was reset.

By replacing the static fields of `ScriptEngineSelector` with a new class `InternalScriptEngineSelector`, I have almost got it to work.

The remaining issue is with the bindings. To give the script access to `sensors`, `turnouts` and so on. That currently doesn't work for me, but I will look into that.

bobjacobsen reviewed (approved):
Looks good!

dsand47 commented:
Script Entry window, selected Jython Files, entered:

print 'hello'

and selected Execute.

```
14:35:34,828 PanelPro.PanelPro                     INFO  - Main initialization done [main]
Warning: Nashorn engine is planned to be removed from a future JDK release
14:36:26,455 ptionhandler.UncaughtExceptionHandler ERROR - Uncaught Exception caught by jmri.util.exceptionhandler.UncaughtExceptionHandler [AWT-EventQueue-0]
java.lang.NullPointerException: scriptEngineSelector.getSelectedEngine() returns null
	at jmri.script.swing.InputWindow.buttonPressed(InputWindow.java:286)
	at jmri.script.swing.InputWindow.lambda$new$1(InputWindow.java:126)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
```

danielb987 commented:
@dsand47 
I'm unable to reproduce this error. I assume this works fine on master?

danielb987 commented:
PR marked as WIP until the issue found by @dsand47 is resolved.

dsand47 commented:
> @dsand47
> I'm unable to reproduce this error. I assume this works fine on master?

If the the user-interface.xml has an old entry:

`        <comboBox xmlns="" lastSelected="Jython Files" name="jmri.script.swing.InputWindow.language"/>
`

When `Script Entry` is selected, the selection combo box is blank (most likely the index is -1):

![Screen Shot 2022-05-16 at 5 06 33 PM](https://user-images.githubusercontent.com/17280992/168690924-100030a2-2abc-4d04-9a60-4a2a3c7131ab.png)

Even after selecting the correct combo box entry, the NPE occurs.

The new entry is:

`        <comboBox xmlns="" lastSelected="python" name="jmri.script.swing.InputWindow.language"/>
`

If there is no user-interface.xml entry, then it defaults to Jython Files and works properly.

danielb987 commented:
Thanks! I have added a check for null when JMRI tries to set the language.

danielb987 commented:
@dsand47 
I have fixed the issue you found. Can you take a new look?

dsand47 commented:
> @dsand47
> I have fixed the issue you found. Can you take a new look?

It is partially fixed.

When switching from 4.99.8, the correct entry is selected, but there is a warning:

`[java] 11:02:37,534 script.ScriptEngineSelector           WARN  - Cannot select engine for the language Jython Files [AWT-EventQueue-0]`

The entered script does run ok.

The entry in user-interface.xml does not get updated until ECMA... is selected and then back to Jython Files.

danielb987 commented:
> When switching from 4.99.8, the correct entry is selected, but there is a warning:
> 
> The entry in user-interface.xml does not get updated until ECMA... is selected and then back to Jython Files.

Both are now fixed. I have added code for backwards compability so that the warning is not shown for pre 4.99.9 setting. This also means that the previous selection (Jython/EcmaScript) the user has is preserved but converted to the new style. And this is stored immediately.

dsand47 reviewed (approved):
Works good now.

Going backwards will default back to ECMAScript but that is expected.
